pkg/localstorev2/reserve.go

```diff
@@ -0,0 +1,484 @@
+	return db.reserve.Get(ctx, db.repo, addr)
+}
+
+func (db *DB) ReserveHas(addr swarm.Address) (bool, error) {
+	return db.reserve.Has(db.repo.IndexStore(), addr)
+}
+
+// ReservePutter returns a PutterSession for inserting chunks into the reserve.
+func (db *DB) ReservePutter(ctx context.Context) PutterSession {
+
+	trx, commit, rollback := db.repo.NewTx(ctx)
+	reservePutter := db.reserve.Putter(trx)
+
+	pos := make(map[uint8]bool)
+	count := 0
+
+	// lock to avoid Putting a chunk that expires during the session.
+	db.lock.Lock(reserveLock)
+
+	return &putterSession{
+		Putter: storage.PutterFunc(func(ctx context.Context, chunk swarm.Chunk) error {
+			err := reservePutter.Put(ctx, chunk)
+			if err != nil {
+				return err
+			}
+			pos[db.po(chunk.Address())] = true
+			count++
+			return nil
+		}),
+		done: func(swarm.Address) error {
+			defer db.lock.Unlock(reserveLock)
+			err := commit()
+			if err != nil {
+				return err
+			}
+			db.reserve.AddSize(count)
+			for po := range pos {
+				db.reserveBinEvents.Trigger(string(po))
+			}
+			if !db.reserve.IsWithinCapacity() {
+				db.events.Trigger(reserveOverCapacity)
+			}
+			return nil
+		},
+		cleanup: func() error {
+			defer db.lock.Unlock(reserveLock)
+			return rollback()
+		},
+	}
+}
+
+// EvictBatch evicts all chunks belonging to a batch from the reserve.
+func (db *DB) EvictBatch(ctx context.Context, batchID []byte) error {
+	return db.evictBatch(ctx, batchID, swarm.MaxBins)
+}
+
+func (db *DB) evictBatch(ctx context.Context, batchID []byte, bin uint8) error {
+
+	db.lock.Lock(reserveLock)
+	defer db.lock.Unlock(reserveLock)
+
+	for b := uint8(0); b < bin; b++ {
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		txnRepo, commit, rollback := db.repo.NewTx(ctx)
+
+		evicted, err := db.reserve.EvictBatchBin(txnRepo, batchID, b)
+		if err != nil {
+			_ = rollback()
+			return err
+		}
+
+		err = commit()
+		if err != nil {
+			return err
+		}
+
+		db.reserve.AddSize(-evicted)
+	}
+
+	return nil
+}
+
+func (db *DB) unreserve(ctx context.Context) error {
+
+	withinCap := false
+	radius := db.reserve.Radius()
+	defer db.events.Trigger(reserveUnreserved)
+
+	for {
+
+		err := db.batchstore.Iterate(func(b *postage.Batch) (bool, error) {
+
+			err := db.evictBatch(ctx, b.ID, radius)
+			if err != nil {
+				return false, err
+			}
+
+			if db.reserve.IsWithinCapacity() {
+				withinCap = true
+				return true, nil
+			}
+
+			return false, nil
+		})
+		if err != nil {
+			return err
+		}
+		if withinCap {
+			return nil
+		}
+
+		radius++
+		_ = db.reserve.SetRadius(db.repo.IndexStore(), radius)
+	}
+}
+
+// ReserveLastBinIDs returns all of the highest binIDs from all the bins in the reserve.
+func (db *DB) ReserveLastBinIDs() ([]uint64, error) {
+	return db.reserve.LastBinIDs(db.repo.IndexStore())
+}
+
+// BinC is the result returned from the SubscribeBin channel that contains the chunk address and the binID
+type BinC struct {
```

Comment:
I think its better to just return the constructed chunk. This way pullsync can just check for stamp validity and directly send it.

I dont think we should expose this BinID outside the package. This will limit us if we come up with different design later. The user of this `SubscribeBin` function only cares about the chunks and not the internal data structures in the implementation.

Reply:
the steps of the pullsync are:

peer1 requests chunks with a certain binID interval from peer2
peer2 prepares an offer of addresses and sends it
peer1 picks the addresses he wants from that offer and sends back the list of addresses
peer2 gets the chunks from the reserve and sends them to peer1

basically peer2 wants to iterate on the list of addresses and NOT chunks

Reply:
with the multi stamp per chunk changes, we may need to rewrite the pullsync